packages/runtime/runtime-utils/src/dataStoreHelpers.ts

```diff
@@ -36,29 +39,35 @@ export function exceptionToResponse(err: any): IResponse {
         };
     }
 
-    // Capture error objects, not stack itself, as stack retrieval is very expensive operation, so we delay it
-    const errWithStack = generateErrorWithStack();
-
+    const normalizedError = normalizeError(err);
     return {
         mimeType: "text/plain",
-        status,
-        value: `${err}`,
-        get stack() { return ((err?.stack) as (string | undefined)) ?? errWithStack.stack; },
+        status: 500,
+        value: normalizedError.message,
+        get stack() { return normalizedError.stack; },
     };
 }
 
-export function responseToException(response: IResponse, request: IRequest): Error {
-    const message = response.value;
-    const errWithStack = generateErrorWithStack();
-    const responseErr: Error & IResponseException = {
-        errorFromRequestFluidObject: true,
-        message,
-        name: "Error",
-        code: response.status,
-        get stack() { return response.stack ?? errWithStack.stack; },
-    };
+/**
+ * Error class for error instances generated from an error IResponse
+ */
+class ResponseException extends LoggingError implements IResponseException {
+    readonly errorFromRequestFluidObject: true = true;
+    readonly code: number;
+    constructor(response: IResponse) {
+        const responseValue = response.value;
+        const message = typeof responseValue === "string" ? responseValue : String(responseValue);
+        super(message);
+
+        this.code = response.status;
+        if (response.stack !== undefined) {
+            this.overwriteStack(response.stack);
+        }
+    }
+}
 
-    return responseErr;
+export function responseToException(response: IResponse, request: IRequest): Error {
```

Comment:
Would it be useful to expose the extra properties in a type-safe manner? (double check me on the syntax if so).
```suggestion
export function responseToException(response: IResponse, request: IRequest): Error & IResponseException {
```